### ISSUE TITLE
Write scripts in Python?

### ISSUE THREAD

**Issue by illwieckz** [Unvanquished/Unvanquished#2200]
[Here](https://github.com/Unvanquished/release-scripts/pull/24#issuecomment-1242820823) @slipher wrote, as I was extending the `build-release` scripts:

> Being opposed to complex shell scripts […]

And in a recent past when I [tried to make the script to build deps](https://github.com/DaemonEngine/Daemon/pull/561) more robust by making it self-aware of what it was doing (and then becoming able to check what it was doing) this effort turned into a disaster because I faced the wall of the belief that a bash program should not become long or complex.

I get the feeling that there is a belief there exist two kind of programs, programs that can do anything if not written in bash, and programs in bash that are not real programs and that should not grow.

To me writing a program in bash language is just the choice of a language among others. I myself often decide to use bash among other choices because 1. it is interpreted, 2. bash is almost ubiquitous (if I restrict to the old bash shipped by default in macos). So I can start long, large, complex projects deciding to go the bash way, never purposing those projects to not be complex but the exact opposite. I can chose bash as an implementation detail _after_ I decide to write something expected to be complex, and I can chose bash in order to write such complex things.

So, at some point, all my attempts to write utilities in a interpreted language to make things that are a bit complex will face opposition if I chose bash among other available languages. And more likely this kind of opposition will be faced years after the utility have been written, when adding this or that over the existing code.

[Here](https://github.com/DaemonEngine/Daemon/pull/561#issuecomment-1138761080) I wrote (in deps rewriting thread):

> Note: I would not be against porting this to another language like python in the future to get rid of the bash wizardry.

Though, python isn't as good as bash and other language with direct access to shell commands in general when it's needed to run a bunch of shell commands. It's even hard to compete with the efficiency of `xargs` when it comes to process in parallel and with efficient job management a bunch of inputs you may even don't know in advance and that may even been feed on the fly.

Some may say that if someone needs more than bash, perl is the best candidate, but unfortunately no one writes perl today, including myself.

One strong argument in favor of Python is that we already require Python for multiple things in the game, so Python is already ubiquitous. Any people who is expected to build the game is expected to already have a working Python installation. So we can unlock full python abilities (and rich and robust data structures like dictionaries that are not available in old bash on macos), and even when restricting to the default python modules we get a lot of features.

So, since Python is considered ubiquitous, why not writing our various utilities in python? I mean those that are expected to be run in development process: building the deps, making a release build, etc.

**Comment by illwieckz** [comment]
A side topic is that by acknowledging Python is required to build the game, we may make cmake delegate more stuff to Python.

For example some people may want to see a cmake target to build a basic dpk. Cmake may be even worst than bash to make a dpk packager, so if we assume Python is there, we can write some python code expected to be called by CMake.

We already do that for CBSE and NaCl if I'm right.

**Comment by slipher** [comment]
I opened #2201 for a start at getting rid of shell scripts :)

>  bash is almost ubiquitous

But not entirely, I can have a Bash-free environment when developing with Visual Studio.

To expand on the shell hate a bit, it is the C of scripting languages. You have the opportunity for huge bugs/security vulnerabilities on every line (compare buffer overflows and quoting bugs). Basic programming facilities are missing and you have to roll them yourself. So writing complex Bash programs in 2022 should get you yelled at the same way as writing new C programs.

I think there is actually a good case for using shell with external_deps/build.sh since it is shelling out to other programs 90% of the time. Perhaps build-release also.

Of the shell scripts I know about, download-paks seems like the best candidate for converting to Python. The logic is a bit complex and it doesn't shell out to so many other programs. It's been buggy when I tried to use it (on Mac). And it could be useful for Windows users too.

**Comment by illwieckz** [comment]
> But not entirely, I can have a Bash-free environment when developing with Visual Studio.

Good point! I always forget about Visual Studio (I never seen one in my life 😁️)!

> To expand on the shell hate a bit, it is the C of scripting languages.

So now we come from shell hate to C hate. 😅️ I'm such a dinosaur. In fact no, dinosaurs get a lot of love and attention (except they're dead 😄️). What about C shell then? I'm joking, even myself don't use that. 🙊️

> Of the shell scripts I know about, download-paks seems like the best candidate for converting to Python.

I agree, at some point I even suggested to make an headless variant of the updater to replace this. 👀️

And using a more robust language like Python would even allow to implement “downloading with curl or wget from multiple CDN server at once”.

On a related topic, I propose to use a folder named `utils/`. Here with Unvanquished repository this would contain:

- `appveyor/`, we may even externalize/mutualize some CI scripts from their `.yml` files,
- `download-paks`,
- `dist/scripts/record-benchmark-demo`,  
  the other scripts from `dist/scripts` are not meant to be run by us but to be packaged to be run by others, they're kind of extra assets for distro packagers, so them in `dist` is find. I would even rename `dist` to `samples` at some point, I only named `dist` dist` because I was looking for the `sample` word without finding it at the time;
- `src/utils/`  
  we can even move other scripts to this one.
- `build-release/` (unless we move them to Dæmon repository directly).

**Comment by slipher** [comment]
Opened a new thread #2202 for layout bikeshedding.

**Comment by slipher** [comment]
Even something as simple as update-version-number might benefit. With a more featureful language, we could
- Avoid creating (and leaking) temp files
- Raise an error if one of the substitutions fails to match anything
- Store the changes in memory and only write out at the end if all replacements are successful
- Have a dry-run mode

If someone tried to add all this in Bash I would complain about overengineering of course.

**Comment by illwieckz** [comment]
Good idea.

In fact there is no need for temp files in bash (at least I'm sure about text files) but even myself use tempfiles for convenience. 😄️

**Comment by bmorel** [comment]
> To expand on the shell hate a bit, it is the C of scripting languages.

If switching lang, I'd say a minimum would be to switch for a tech that have static analysis which actually works, instead of just taking the next lang that is fashionable.
For that, a start is to have a static type system imo, and that the interpreter would check the whole script for syntax errors before even trying to run it. This might be slow, but in those cases, is performance really a problem?
Unlike C, which have compilers and linters, bourne shell lack any of (usable) tools.

Python? Why not, but maybe start by checking for the presence of such tools before adopting it, or the history will be repeat.

**Comment by illwieckz** [comment]
> bourne shell lack any of (usable) tools.

shellcheck isn't bad (I should use it more)

> If switching lang, I'd say a minimum would be to switch for a tech that have […]

Something between bad and good is always better than bad anyway, and we already require developers to install Python, which is a strong argument for it.

**Comment by bmorel** [comment]
> shellcheck isn't bad (I should use it more)

Each time I used it, it's "advices" led me to write more buggy code.

> 
bourne shell lack any of (usable) tools.

shellcheck isn't bad (I should use it more)

If switching lang, I'd say a minimum would be to switch for a tech that have […]

> Something between bad and good is always better than bad anyway, and we already require developers to install Python, which is a strong argument for it.

This may be right, but the question is, why do we need python right now? There are 2 reasons: CBSE C++ code generator being written in python and relying on yaml, and NaCl.
NaCl will, hopefully, disappear one day. This is the reason why we actually require python 2 after all.
As for the CBSE, I don't really think it's a good reason to use python. I usually despise code code generators, and it's even more true when they are used to generate C++ templates relying on variadic templates, for a very simple reason: modern C++ *can* implement DSL at build time, as several boost libs shows. The python requirement to build system this adds, compared to the problems CBSE brings, is, in my opinion (which is not shared, I know), not worth it.

You will note that my arguments are *not* against python itself.
I am concerned about the code's quality, and about ways to automate requirements for a higher code quality, so that contributor's time can be spent on more valuable stuff than nitpicking.
Taking (again) the example of the CBSE, which is the only tool we have supporting python3, it fails at giving clear error messages when the build fails for one reason or another. It is, also, undocumented and unmaintained.
To me, this stuff being written in python is not a good reason to write more python. If I was about to generalize, I would even say it's a good reason to *avoid* python, and I would say it's the average quality of python tools. I do think it, but I still do hope I'm wrong, and I know by experience that a language or technology is never responsible for mis-uses.

My answer was more about inviting you guys to think about the requirements and the tools around before deciding for a tech. I was only inviting you to think forward, instead of jumping in a lang which might not be tailored for the uses you intend to use it for.
This brings one question: what kind of tasks are those scripts excepted to do? If it's for system, then I wonder if python is really good at it. So far, I only have heard about 3 tools for system system stuff that have rather wide adoption: bourne shell, powershell and go. No preference order there. One of them is not a script lang, though.
